tests/test_ecephys/test_tools_spikeinterface.py

```diff
@@ -1009,6 +1028,173 @@ def test_write_units_table_in_processing_module(self):
+        # test that waveforms and stds are the same
+        unit_ids = we.unit_ids
+        for unit_index, _ in enumerate(nwbfile.units.id):
+            wf_mean_si = we.get_template(unit_ids[unit_index])
+            wf_mean_nwb = nwbfile.units[unit_index]["waveform_mean"].values[0]
+            np.testing.assert_array_almost_equal(wf_mean_si, wf_mean_nwb)
+            wf_sd_si = we.get_template(unit_ids[unit_index], mode="std")
+            wf_sd_nwb = nwbfile.units[unit_index]["waveform_sd"].values[0]
+            np.testing.assert_array_almost_equal(wf_sd_si, wf_sd_nwb)
+
+    def test_write_single_segment(self):
+        """This test that the waveforms are written appropriately for the single segment case"""
+        write_waveforms(waveform_extractor=self.single_segment_we, nwbfile=self.nwbfile1, write_electrical_series=True)
+        self._test_waveform_write(self.single_segment_we, self.nwbfile1)
+        self.assertIn("ElectricalSeriesRaw", self.nwbfile1.acquisition)
+
+    def test_write_multiple_segments(self):
+        """This test that the waveforms are written appropriately for the multi segment case"""
+        write_waveforms(waveform_extractor=self.multi_segment_we, nwbfile=self.nwbfile2, write_electrical_series=False)
+        self._test_waveform_write(self.multi_segment_we, self.nwbfile2)
+
+    def test_write_subset_units(self):
+        """This test that the waveforms are sliced properly based on unit_ids"""
+        subset_unit_ids = self.single_segment_we.unit_ids[::2]
+        write_waveforms(waveform_extractor=self.single_segment_we, nwbfile=self.nwbfile3, unit_ids=subset_unit_ids)
+        self._test_waveform_write(self.we_slice, self.nwbfile3, test_properties=False)
+
+        self.assertEqual(len(self.nwbfile3.units), len(subset_unit_ids))
+        self.assertTrue(all(str(unit_id) in self.nwbfile3.units["unit_name"][:] for unit_id in subset_unit_ids))
+
+    def test_write_recordingless(self):
+        """This test that the waveforms are sliced properly based on unit_ids"""
+        write_waveforms(
+            waveform_extractor=self.we_recless,
+            nwbfile=self.nwbfile4,
+            recording=self.we_recless_recording,
+            write_electrical_series=True,
+        )
+        self._test_waveform_write(self.we_recless, self.nwbfile4, test_properties=False)
+
+        # check that not passing the recording raises and Exception
+        with self.assertRaises(Exception) as context:
+            write_waveforms(
+                waveform_extractor=self.we_recless,
+                nwbfile=self.nwbfile4,
+                recording=self.we_recless_recording,
+                write_electrical_series=True,
+            )
+
+    def test_cleanup_tmp_properties(self):
+        # test that if write_waveforms fails, temporary properties are cleaned up
+        try:
+            write_waveforms(waveform_extractor=self.we_recless, nwbfile=self.nwbfile5, recording=None)
+        except:
```

Comment:
it does! The `write_waveforms` will trigger an exception because:
1. Waveform extractor is recordingless
2. `recording` is `None`

In that case we check that temporary properties are properly cleaned up

Reply:
I know, but as a test it can pass 'succesfully' if, due to a potential future bug, that exception does not get raised, in which case no assertions on the cleanup are ever made. If you use `finally`, the cleanup is asserted independently of whether or not the exception raise itself was successful.

An even stronger test, if you want to keep this structure, would be to alias the exception like

```python
exception Exception as exc:
    exception_type = type(exc)
    exception_message = str(exc)
    ...
```

and then outside the `try/except` block, also assert that the exception was raised as you intended. Even with that approach, I'd still recommend

```python
try:
    # write waveforms with intentionally forced error
except Exception as exc:
    # grab exception details
finally:
    # assert cleanup - important thing is this line will ALWAYS run
#assert exception details match expectation
```

Reply:
Although, is this even still necessary since it seems as if you've returned to using a copy of the object rather than temporary properties?